W-13791369 (hot deploy) and W-13932478 revisions debug cont duke (#125)
mulesoft/docs-code-builder: modules/ROOT/pages/debug-a-mule-application.adoc

image::debugger-view.png["Elements of the debugger view"]

[calloutlist]

> . *Start Debugging*: Start an instance of a Mule application in debugging mode.

Reviewer: I believe you may need the 's' for all these. As your kind of saying, Start Debugging starts an instance. I feel like without it, it implies a command to the user or that Start Debugging is plural. Just my opinion though.

Reviewer: Maybe they are commands though, but I think this is a callout list.

Author: I think I was trying to keep the descriptions parallel so that I could use "Navigate" in another description. I'm not sure if it creates confusion, tbh. It's a good point to consider. I've got a release soon and need to get this out, so I'll keep this one as-is for now.

> . *Start Debugging*: Start an instance of a Mule application in debugging mode.

Author: ```suggestion
. *Start Debugging*: Starts an instance of a Mule application in debugging mode
```

> . *Start Debugging*: Starts an instance of a Mule application in debugging mode

Author: ```suggestion
. *Start Debugging*: Start an instance of a Mule application in debugging mode.
```

Author: I think I was trying to keep the descriptions parallel so that I could use "Navigate" in another description

> . *Debug Toolbar*: Provides shortcuts to control debugging of Mule applications:

Author: @greobasco Simplified Debug Toolbar to avoid repeating details.

> . *Debug Toolbar*: Provides shortcuts to control debugging of Mule applications:

Author: ```suggestion
. *Debug Toolbar*: Use shortcuts to control debugging of Mule applications:
```

> . *Debug Toolbar*: Pause execution of a running application at elements in the configuration XML. Restart and stop the application.

Reviewer 2: I think we should keep the word "breakpoints" in this description as it will only pause if a breakpoint has been set.

Maybe like this:
Provides shortcuts to control debugging of Mule applications. It allows to navigate through elements in the configuration XML when a breakpoint has been set. It also provides controls to restart or stop the running application.

> . *Debug Toolbar*: Pause execution of a running application at elements in the configuration XML. Restart and stop the application.

Author: ```suggestion
. *Debug Toolbar*: Provides shortcuts to control debugging of Mule applications:
```

+
image::debugger-toolbar.png[width=250]
+
[arabic]

> .. *Continue*: Proceed to the next breakpoint, if present, and pause execution of the application. Otherwise, continue to the end of the application. Key command: F5

Author: @greobasco Are you ok with this revision based on your suggestion?

> .. *Continue*: Proceed to the next breakpoint, if present, and pause execution or the execution of the application. Otherwise, continue to the end of the application. Key command: F5

Author: ```suggestion
.. *Continue*: Proceed to the next breakpoint, if present, and pause execution of the application. Otherwise, continue to the end of the application. Key command: F5
```

> .. *Continue*: Proceed to the end of the application flow or pauses execution at Flow Reference component (`<flow-ref/>`) within the flow but does not proceed to another flow. Key command: F5

Reviewer 2: I would change this as the Continue button does not relate to Flow ref.

.. *Continue*: Proceeds the execution of the application flow. If there is a breakpoint set anywhere after the continuing element, the execution will be paused at that point. Key command: F5

Author: @greobasco I notice that there is no situation where we just proceed to the next breakpoint. We always go to the next component regardless of whether there is a breakpoint or not.

Author: @greobasco Since the buttons are described below, I'lm just shortening to say that the toolbar helps with debugging instead of repeating info here.

> .. *Continue*: Proceed to the end of the application flow or pauses execution at Flow Reference component (`<flow-ref/>`) within the flow but does not proceed to another flow. Key command: F5

Author: ```suggestion
.. *Continue*: Proceed to the next breakpoint, if present, and pause execution or the execution of the application. Otherwise, continue to the end of the application. Key command: F5
```

.. *Step Over*: Proceed to the next component in the flow, where execution pauses. At a Flow Ref component (`<flow-ref/>`), clicking *Step Over* proceeds to the next component in the flow but _does not_ follow the Flow Ref logic to a referenced flow or proceed to another flow. Key command: F10
+
.Configuration XML example:
[source,xml]
----
<flow name="flow1" doc:name="Flow" >
  <http:listener path="greet" config-ref="HTTP_Listener_config" doc:name="Listener"/>
  <flow-ref name="myNewFlow" /> <!--1-->
  <logger doc:name="Logger1" message="test1"/> <!--2-->
</flow>

<flow name="flow2" doc:name="Flow" >
  <logger doc:name="Logger2" message="test2"/>
</flow>
----
.*Step Over* behavior:
[calloutlist]
... Skips `<flow-ref/>` logic

> ... Reaches `<logger/>`

Author: ```suggestion
... Reaches `<logger/>` named `Logger1`
```

> ... Reaches `<logger/>` named `Logger1`

Author: @greobasco revised to clarify which logger. In the rendered HTML, there is an association between the `<!--2-->` and a numbered item (2), so it's more clear there but less clear in the Asciidoc source.

> ... Reaches `<logger/>`

Reviewer 2: ... Reaches `Logger1`

Just to make it clearer

> .. *Step Into*: Proceed to the next component in the flow, where execution pauses. At a Flow Ref component (`<flow-ref/>`), clicking *Step Into* follows the logic of the component and proceeds to the first component in the referenced flow. If a component in a flow follows a Flow Reference component, *Step Into* returns to that component after the last component in the referenced flow (or flows). Key command: F11

Reviewer: You use both "Flow Ref component" and "Flow Reference component."

Author: fixed, thx
The name is different in Studio ¯\_(ツ)_/¯

> .. *Step Into*: Proceed to the next component in the flow, where execution pauses. At a Flow Ref component (`<flow-ref/>`), clicking *Step Into* follows the logic of the component and proceeds to the first component in the referenced flow. If a component in a flow follows a Flow Reference component, *Step Into* returns to that component after the last component in the referenced flow (or flows). Key command: F11

Author: ```suggestion
.. *Step Into*: Proceed to the next component in the flow, where execution pauses. At a Flow Ref component (`<flow-ref/>`), clicking *Step Into* follows the logic of the component and proceeds to the first component in the referenced flow. If a component in a flow follows a Flow Ref component, *Step Into* returns to that component after the last component in the referenced flow (or flows). Key command: F11
```

+
.Configuration XML example:
[source,xml]
----
<flow name="flow1" doc:name="Flow" >
  <http:listener path="greet" config-ref="HTTP_Listener_config" doc:name="Listener"/>
  <flow-ref name="myNewFlow" /> <!--1-->
  <logger doc:name="Logger1" message="test1"/> <!--3-->
</flow>

<flow name="myNewFlow" doc:name="Flow" >
  <logger doc:name="Logger2" message="test2"/> <!--2-->
</flow>
----
.*Step Into* behavior:
[calloutlist]
... Enters `<flow-ref/>` to follow logic to `myNewFlow`
... Reaches `Logger2` in next flow, `myNewFlow`
... Returns to flow `flow1` to reach `Logger1`

> .. *Step Out*: Exit the current flow and attempts to return to a previous flow. If a previous flow doesn’t exist, the debugger behaves like *Continue*. Key command: Shift+F11

Reviewer: ```suggestion
.. *Step Out*: Exit the current flow and attempts to return to a previous flow. If a previous flow doesn’t exist, the debugger behaves like *Continue*. Key command: `Shift+F11`
```

Maybe?

Author: Style guideline is to not use a different font face. ¯\_(ツ)_/¯

.. *Restart*: Stop and restart the application without packaging new dependencies, such as new connector dependencies in your `pom.xml` file. Key command: Shift+Cmd+F5 (Mac) or Shift+Ctrl+F5 (Windows)
.. *Stop*: Stop execution of the application. The next debugging session that you start packages the application with any new dependencies. Key command: Shift+F5
. *Variables* panel: Provides Mule event and parameter data. The Mule event includes message payload and attribute details, as well as Mule variable data. The parameters include component data, such as the component name.
. *Run and Debug* (MuleSoft) icon: Opens the *Run and Debug* panel, where you can click *Start Debugging* and find debugging information. Key command: Cmd+Shift+d (Mac) or Ctrl+Shift+d (Windows) 
. *Watch* panel: Helps you evaluate variables and expressions.
. *Call Stack* panel: Identifies the functions or procedures that are currently in the stack.
. *Breakpoints* panel: Lists all breakpoints in your Mule application.

> . *Terminal* console: Displays output of the Maven build process and the results of the deployment to the embedded Mule Runtime engine.

Reviewer: ```suggestion
. *Terminal* console: Displays the output of the Maven build process and the results of the deployment to the embedded Mule Runtime engine.
```

> . *Breakpoint* icon: Identifies a line in the configuration XML to pause execution of the flow so that you can check for and fix any issues

Reviewer: ```suggestion
. *Breakpoint* icon: Identifies a line in the configuration XML to pause execution of the flow so that you can check for and fix any issues.
```

As the above items have periods

> Start by xref:debug-add-logger-set-variables.adoc[Adding Components to Help You Debug Your Application].

Reviewer: ```suggestion
Start by xref:debug-add-logger-set-variables.adoc[Adding Components to Help You Debug Your Application]:
```

> Hot deploy as you debug your application. A hot deployment updates a running instance of your application but does not package new dependencies, such as new connector dependencies in your `pom.xml`.

Reviewer: ```suggestion
Hot deploy as you debug your application. A hot deployment updates a running instance of your application but does not package new dependencies, such as new connector dependencies in your `pom.xml`:
```

> . Make sure your Mule application is running in the IDE.

Reviewer: ```suggestion
. Ensure your Mule application is running in the IDE.
```
